node/src/client/mod.rs

```diff
@@ -35,21 +35,58 @@ use snarkvm::{
     console::network::Network,
     ledger::{
         block::{Block, Header},
-        puzzle::{Puzzle, Solution},
+        puzzle::{Puzzle, Solution, SolutionID},
         store::ConsensusStorage,
         Ledger,
     },
+    prelude::block::Transaction,
 };
 
 use aleo_std::StorageMode;
 use anyhow::Result;
 use core::future::Future;
+use lru::LruCache;
 use parking_lot::Mutex;
 use std::{
     net::SocketAddr,
-    sync::{atomic::AtomicBool, Arc},
+    num::NonZeroUsize,
+    sync::{
+        atomic::{
+            AtomicBool,
+            AtomicUsize,
+            Ordering::{Acquire, Relaxed},
+        },
+        Arc,
+    },
+    time::Duration,
 };
-use tokio::task::JoinHandle;
+use tokio::{task::JoinHandle, time::sleep};
+
+/// The maximum number of deployments to verify in parallel.
+/// Note: worst case memory to verify a deployment (MAX_DEPLOYMENT_CONSTRAINTS = 1 << 20) is ~2 GiB.
+const MAX_PARALLEL_DEPLOY_VERIFICATIONS: usize = 5;
+/// The maximum number of executions to verify in parallel.
+/// Note: worst case memory to verify an execution is 0.01 GiB.
+const MAX_PARALLEL_EXECUTE_VERIFICATIONS: usize = 1000;
```

Comment:
the default (and the one we're currently using) [blocking task](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html) count limit is 512; we might need to raise it now, especially considering that there are other users of blocking tasks all around the codebase

Reply:
I will run another benchmark to see how fast executions can be verified, perhaps its just a few ms.

From how many ms would you recommend using blocking tasks? Also could you make a recommendation for the new count limit, assuming scenario's 1. executions are verified using spawns 2. executions are verified using spawn_blocking?

Reply:
For anything that goes into the millisecond scale we should be using blocking tasks, otherwise we will negatively affect the async executor. In addition, for any operations that aren't just slow but are heavy on the CPU we should be using `rayon` to execute them _in addition to_ using blocking tasks ([example](https://github.com/AleoNet/snarkOS/commit/a4c905e2e697b7abeebe3aa16c6fbc6e4591c510)).

As for bumping the number of blocking threads... I would say it could be a lot larger (especially if the actual work is offloaded to `rayon` and they are used only to be able to `await` such operations), but we should test how it impacts node memory use etc.

Reply:
As discussed elsewhere: ["After reaching the upper limit, the tasks are put in a queue."](https://docs.rs/tokio/latest/tokio/task/fn.spawn_blocking.html). Given that changing the task limit may have complex side-effects, I'll mark it as a future TODO for further optimization